File: DerpCode.API/Models/Entities/UserPreferences.cs
using System;

namespace DerpCode.API.Models.Entities;

public sealed class UserPreferences : IIdentifiable<int>, IOwnedByUser<int>
{
    public int Id { get; set; }

    public int UserId { get; set; }

    public User User { get; set; } = default!;

    public Preferences Preferences { get; set; } = new();

    public DateTimeOffset LastUpdated { get; set; }
}

public sealed class Preferences
{
    public UserUIPreference UIPreference { get; set; } = new();

    public UserCodePreference CodePreference { get; set; } = new();

    public UserEditorPreference EditorPreference { get; set; } = new();
}

public sealed class UserEditorPreference
{
    public bool EnableFlameEffects { get; set; } = true;
}

public sealed class UserCodePreference
{
    public LanguageType DefaultLanguage { get; set; } = LanguageType.JavaScript;
}

public sealed class UserUIPreference
{
    public UITheme UITheme { get; set; } = UITheme.Dark;

    public int PageSize { get; set; } = 5;
}

public enum UITheme
{
    Dark,
    Light,
    Custom
}

File: derpcode-ui/src/types/userPreferences.ts
import type { Language } from './models';

export enum UITheme {
  Dark = 'Dark',
  Light = 'Light',
  Custom = 'Custom'
}

export interface UserEditorPreferenceDto {
  enableFlameEffects: boolean;
}

export interface UserCodePreferenceDto {
  defaultLanguage: Language;
}

export interface UserUIPreferenceDto {
  uiTheme: UITheme;
  pageSize: number;
}

export interface PreferencesDto {
  uiPreference: UserUIPreferenceDto;
  codePreference: UserCodePreferenceDto;
  editorPreference: UserEditorPreferenceDto;
}

export interface UserPreferencesDto {
  id: number;
  userId: number;
  lastUpdated: string;
  preferences: PreferencesDto;
}
